cudax/include/cuda/experimental/__memory_resource/any_resource2.cuh

```diff
@@ -0,0 +1,188 @@
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+// If the memory resource header was included without the experimental flag,
+// tell the user to define the experimental flag.
+#if defined(_CUDA_MEMORY_RESOURCE) && !defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE)
+#  error "To use the experimental memory resource, define LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE"
+#endif
+
+// cuda::mr is unavable on MSVC 2017
+#if defined(_CCCL_COMPILER_MSVC_2017)
+#  error "The any_resource header is not supported on MSVC 2017"
+#endif
+
+#if !defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE)
+#  define LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE
+#endif
+
+#include <cuda/__memory_resource/get_property.h>
+#include <cuda/__memory_resource/properties.h>
+#include <cuda/__memory_resource/resource.h>
+#include <cuda/std/__concepts/__concept_macros.h>
+#include <cuda/std/__utility/forward.h>
+#include <cuda/std/optional>
+
+#include <cuda/experimental/__utility/basic_any.cuh>
+
+namespace cuda::experimental::mr
+{
+using __property_value_or_void _CCCL_NODEBUG_ALIAS = //
+  _CUDA_VSTD::__type_try_catch< //
+    _CUDA_VSTD::__type_indirect_quote<__property_value_t>, // try to get the property value
+    _CUDA_VSTD::__type_always<void>>; // if it fails, return void
+
+template <class _Property>
+using __property_result_t _CCCL_NODEBUG_ALIAS = _CUDA_VSTD::__type_call1<__property_value_or_void, _Property>;
+
+template <class _Property>
+struct __with_property
+{
+  template <class _Ty>
+  _CUDAX_PUBLIC_API static auto __get_property(const _Ty& __obj, _Property __prop) noexcept //
+    -> __property_result_t<_Property>
+  {
+    static_assert(noexcept(get_property(__obj, __prop)));
+    if constexpr (!_CUDA_VSTD::is_same_v<__property_result_t<_Property>, void>)
+    {
+      return get_property(__obj, __prop);
+    }
+  }
+
+  template <class...>
+  struct __iproperty : interface<__iproperty>
+  {
+    _CUDAX_API friend auto get_property(const __iproperty& __obj, _Property __prop) noexcept //
+      -> __property_result_t<_Property>
+    {
+      if constexpr (!_CUDA_VSTD::is_same_v<__property_result_t<_Property>, void>)
+      {
+        return __cudax::virtcall<&__get_property<__iproperty>>(&__obj, __prop);
+      }
+    }
+
+    template <class _Ty>
+    using overrides _CCCL_NODEBUG_ALIAS = overrides_for<_Ty, &__get_property<_Ty>>;
+  };
+};
+
+template <class _Property>
+using __iproperty = typename __with_property<_Property>::template __iproperty<>;
+
+template <class... _Properties>
+using __iproperty_set = iset<__iproperty<_Properties>...>;
+
+template <class...>
+struct __ibasic_resource : interface<__ibasic_resource, extends<icopyable<>, iequality_comparable<>>>
```

Comment:
Does this work if a resource is immovable?

Reply:
no resources need to be copyable